Reduce exception noise

there's a lot of noise when closing nodes and sessions

```
[main] INFO org.mikado.imc.topology.SessionManager - added session: tcp-127.0.0.1:55666->tcp-127.0.0.1:9999
[klava.topology.RegisterNodeCoordinator-1] INFO org.mikado.imc.topology.SessionManager - added session: tcp-127.0.0.1:9999->tcp-127.0.0.1:55666
[main] INFO klava.topology.ClientNode - subscribed to tcp-localhost:9999 as Hello
Hello World
tcp-127.0.0.1:9999->tcp-127.0.0.1:55666: [org.mikado.imc.topology.ProtocolThread-1] ERROR klava.proto.MessageState - protocol error in message state
java.io.EOFException
	at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:297)
	at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275)
	at org.mikado.imc.protocols.IMCUnMarshaler.readStringLine(IMCUnMarshaler.java:75)
	at klava.proto.MessageProtocolLayer.doCreateUnMarshaler(MessageProtocolLayer.java:59)
	at org.mikado.imc.protocols.ProtocolStack.createUnMarshaler(ProtocolStack.java:154)
	at org.mikado.imc.protocols.ProtocolState.createUnMarshaler(ProtocolState.java:145)
	at org.mikado.imc.protocols.ProtocolState.getUnMarshaler(ProtocolState.java:174)
	at klava.proto.MessageState.enter(MessageState.java:215)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:110)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:140)
	at org.mikado.imc.topology.ProtocolThread.execute(ProtocolThread.java:47)
	at org.mikado.imc.topology.NodeProcess.run(NodeProcess.java:101)
tcp-127.0.0.1:55666->tcp-127.0.0.1:9999: [org.mikado.imc.topology.ProtocolThread-2] ERROR klava.proto.MessageState - protocol error in message state
java.net.SocketException: Socket closed
	at java.base/sun.nio.ch.NioSocketImpl.endRead(NioSocketImpl.java:243)
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:323)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:346)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:796)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:1099)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:291)
	at java.base/java.io.BufferedInputStream.implRead(BufferedInputStream.java:325)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:312)
	at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:295)
	at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275)
	at org.mikado.imc.protocols.IMCUnMarshaler.readStringLine(IMCUnMarshaler.java:75)
	at klava.proto.MessageProtocolLayer.doCreateUnMarshaler(MessageProtocolLayer.java:59)
	at org.mikado.imc.protocols.ProtocolStack.createUnMarshaler(ProtocolStack.java:154)
	at org.mikado.imc.protocols.ProtocolState.createUnMarshaler(ProtocolState.java:145)
	at org.mikado.imc.protocols.ProtocolState.getUnMarshaler(ProtocolState.java:174)
	at klava.proto.MessageState.enter(MessageState.java:215)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:108)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:140)
	at org.mikado.imc.topology.ProtocolThread.execute(ProtocolThread.java:47)
	at org.mikado.imc.topology.NodeProcess.run(NodeProcess.java:101)
[klava.topology.RegisterNodeCoordinator-1] ERROR klava.topology.KlavaNodeCoordinator - uncaught exception in coordinator klava.topology.RegisterNodeCoordinator-1
klava.KlavaConnectionException: java.net.SocketException: Socket closed
	at klava.topology.KlavaNode.handleIncoming(KlavaNode.java:432)
	at klava.topology.KlavaNode.register(KlavaNode.java:348)
	at klava.topology.KlavaNodeCoordinatorProxy.register(KlavaNodeCoordinatorProxy.java:99)
	at klava.topology.KlavaNodeCoordinator.register(KlavaNodeCoordinator.java:203)
	at klava.topology.RegisterNodeCoordinator.executeProcess(RegisterNodeCoordinator.java:66)
	at klava.topology.KlavaNodeCoordinator.execute(KlavaNodeCoordinator.java:84)
	at org.mikado.imc.topology.NodeCoordinator.run(NodeCoordinator.java:109)
Caused by: java.net.SocketException: Socket closed
	at org.mikado.imc.protocols.tcp.TcpSessionStarter.accept(TcpSessionStarter.java:159)
	at org.mikado.imc.protocols.ProtocolStack.accept(ProtocolStack.java:341)
	at org.mikado.imc.topology.ConnectionServer.accept(ConnectionServer.java:156)
	at org.mikado.imc.topology.ConnectionServer.accept(ConnectionServer.java:140)
	at org.mikado.imc.topology.Node.accept(Node.java:228)
	at klava.topology.KlavaNode.handleIncoming(KlavaNode.java:401)
	... 6 more
[org.mikado.imc.topology.ProtocolThread-1] INFO org.mikado.imc.topology.SessionManager - removed session: tcp-127.0.0.1:9999->tcp-127.0.0.1:55666
[klava.topology.RegisterNodeCoordinator-1] ERROR org.mikado.imc.topology.NodeCoordinator - uncaught exception in coordinator klava.topology.RegisterNodeCoordinator-1
org.mikado.imc.common.IMCException: Uncaught exception
	at klava.topology.KlavaNodeCoordinator.execute(KlavaNodeCoordinator.java:94)
	at org.mikado.imc.topology.NodeCoordinator.run(NodeCoordinator.java:109)
Caused by: klava.KlavaConnectionException: java.net.SocketException: Socket closed
	at klava.topology.KlavaNode.handleIncoming(KlavaNode.java:432)
	at klava.topology.KlavaNode.register(KlavaNode.java:348)
	at klava.topology.KlavaNodeCoordinatorProxy.register(KlavaNodeCoordinatorProxy.java:99)
	at klava.topology.KlavaNodeCoordinator.register(KlavaNodeCoordinator.java:203)
	at klava.topology.RegisterNodeCoordinator.executeProcess(RegisterNodeCoordinator.java:66)
	at klava.topology.KlavaNodeCoordinator.execute(KlavaNodeCoordinator.java:84)
	... 1 more
Caused by: java.net.SocketException: Socket closed
	at org.mikado.imc.protocols.tcp.TcpSessionStarter.accept(TcpSessionStarter.java:159)
	at org.mikado.imc.protocols.ProtocolStack.accept(ProtocolStack.java:341)
	at org.mikado.imc.topology.ConnectionServer.accept(ConnectionServer.java:156)
	at org.mikado.imc.topology.ConnectionServer.accept(ConnectionServer.java:140)
	at org.mikado.imc.topology.Node.accept(Node.java:228)
	at klava.topology.KlavaNode.handleIncoming(KlavaNode.java:401)
	... 6 more
[org.mikado.imc.topology.ProtocolThread-2] INFO org.mikado.imc.topology.SessionManager - removed session: tcp-127.0.0.1:55666->tcp-127.0.0.1:9999
[org.mikado.imc.topology.ProtocolThread-1] ERROR org.mikado.imc.topology.ProtocolThread - protocol error in thread org.mikado.imc.topology.ProtocolThread-1
java.io.EOFException
	at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:297)
	at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275)
	at org.mikado.imc.protocols.IMCUnMarshaler.readStringLine(IMCUnMarshaler.java:75)
	at klava.proto.MessageProtocolLayer.doCreateUnMarshaler(MessageProtocolLayer.java:59)
	at org.mikado.imc.protocols.ProtocolStack.createUnMarshaler(ProtocolStack.java:154)
	at org.mikado.imc.protocols.ProtocolState.createUnMarshaler(ProtocolState.java:145)
	at org.mikado.imc.protocols.ProtocolState.getUnMarshaler(ProtocolState.java:174)
	at klava.proto.MessageState.enter(MessageState.java:215)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:110)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:140)
	at org.mikado.imc.topology.ProtocolThread.execute(ProtocolThread.java:47)
	at org.mikado.imc.topology.NodeProcess.run(NodeProcess.java:101)
[org.mikado.imc.topology.ProtocolThread-2] ERROR org.mikado.imc.topology.ProtocolThread - protocol error in thread org.mikado.imc.topology.ProtocolThread-2
java.net.SocketException: Socket closed
	at java.base/sun.nio.ch.NioSocketImpl.endRead(NioSocketImpl.java:243)
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:323)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:346)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:796)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:1099)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:291)
	at java.base/java.io.BufferedInputStream.implRead(BufferedInputStream.java:325)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:312)
	at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:295)
	at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275)
	at org.mikado.imc.protocols.IMCUnMarshaler.readStringLine(IMCUnMarshaler.java:75)
	at klava.proto.MessageProtocolLayer.doCreateUnMarshaler(MessageProtocolLayer.java:59)
	at org.mikado.imc.protocols.ProtocolStack.createUnMarshaler(ProtocolStack.java:154)
	at org.mikado.imc.protocols.ProtocolState.createUnMarshaler(ProtocolState.java:145)
	at org.mikado.imc.protocols.ProtocolState.getUnMarshaler(ProtocolState.java:174)
	at klava.proto.MessageState.enter(MessageState.java:215)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:108)
	at org.mikado.imc.protocols.Protocol.start(Protocol.java:140)
	at org.mikado.imc.topology.ProtocolThread.execute(ProtocolThread.java:47)
	at org.mikado.imc.topology.NodeProcess.run(NodeProcess.java:101)
[org.mikado.imc.topology.ProtocolThread-1] INFO org.mikado.imc.topology.NodeProcess - terminated org.mikado.imc.topology.ProtocolThread-1
[org.mikado.imc.topology.ProtocolThread-2] INFO org.mikado.imc.topology.NodeProcess - terminated org.mikado.imc.topology.ProtocolThread-2

```

--- a/CODEGEN_NOTES.md
+++ b/CODEGEN_NOTES.md
@@ -1,0 +1,223 @@
+# XKlaim Code Generation — `done()` and `waitForCompletion()`
+
+This note describes the Java code the XKlaim compiler generates for `node` and `net`
+declarations, and how the `done()` / `waitForCompletion()` lifecycle is meant to work.
+
+---
+
+## What the compiler generates
+
+### For each top-level `node`
+
+A Java class extending `KlavaNode` is produced. It contains:
+
+| Member | Description |
+|---|---|
+| `<NodeName>Process` | Private static inner class extending `KlavaNodeCoordinator`. Its `executeProcess()` body is the DSL node body. |
+| `<locality> = new LogicalLocality(…)` | One `static final` field per environment entry (nodes with `[env]` block). |
+| `setupEnvironment()` | Populates the environment map; only generated for nodes with an `[env]` block. |
+| `addMainProcess()` | Creates the coordinator, registers it via `setMainCoordinator()`, and starts it via `addNodeCoordinator()`. |
+
+Example — DSL input:
+
+```
+package foo
+
+node TestNode [other -> phyloc("localhost:9999")] {
+    out("hello")@other
+    done()
+}
+```
+
+Generated `TestNode.java` (abbreviated):
+
+```java
+public class TestNode extends KlavaNode {
+  private static class TestNodeProcess extends KlavaNodeCoordinator {
+    @Override
+    public void executeProcess() {
+      out("hello", other);
+      done();
+    }
+  }
+
+  private static final LogicalLocality other = new LogicalLocality("other");
+
+  public void setupEnvironment() {
+    addToEnvironment(other, getPhysical(XklaimRuntimeUtil.phyloc("localhost:9999")));
+  }
+
+  public void addMainProcess() throws IMCException {
+    KlavaNodeCoordinator _coordinator = new TestNode.TestNodeProcess();
+    setMainCoordinator(_coordinator);
+    addNodeCoordinator(_coordinator);
+  }
+}
+```
+
+### For each `net`
+
+A Java class extending `LogicalNet` is produced. Each node inside the net becomes
+a public static inner class extending `ClientNode`. The net class also has:
+
+| Member | Description |
+|---|---|
+| `<locality> = new LogicalLocality(…)` | One `static final` field per node (used as the logical locality name). |
+| `<NodeName>` (inner class) | Extends `ClientNode`; same structure as a top-level node but its constructor calls `super(physicalLocality, logicalLocality)`. |
+| constructor | Calls `super(new PhysicalLocality("…"))`. |
+| `addNodes()` | Instantiates all inner nodes, calls `setupEnvironment()` where needed, calls `addManagedNode()` for each, then `addMainProcess()` for each. |
+
+Example — DSL input:
+
+```
+package foo
+
+net HelloNet physical "localhost:9999" {
+    node Hello {
+        out("Hello World")@self
+        in(var String message)@self
+        println(message)
+        done()
+    }
+}
+```
+
+Generated `HelloNet.addNodes()`:
+
+```java
+public void addNodes() throws IMCException {
+    HelloNet.Hello hello = new HelloNet.Hello();
+    addManagedNode(hello);
+    hello.addMainProcess();
+}
+```
+
+### The generated `main` class
+
+One main class per XKlaim source file (named after the file) with a `main(String[])` method that:
+
+1. Instantiates all top-level nodes and nets.
+2. Calls `setupEnvironment()` on nodes that have an environment block.
+3. Calls `addMainProcess()` on each node / `addNodes()` on each net (starts execution).
+4. Calls `waitForCompletion()` on each node and each net (waits for orderly shutdown).
+
+```java
+public static void main(final String[] args) throws Exception {
+    // top-level nodes
+    TestNode testNode = new TestNode();
+    testNode.setupEnvironment();
+    testNode.addMainProcess();
+    testNode.waitForCompletion();
+
+    // nets
+    TestNet testNet = new TestNet();
+    testNet.addNodes();
+    testNet.waitForCompletion();
+}
+```
+
+---
+
+## `done()`
+
+`done()` is defined on `KlavaNodeCoordinator`. Call it at the end of a node body to
+signal that the node has finished its work:
+
+```
+node Hello {
+    out("Hello World")@self
+    in(var String message)@self
+    println(message)
+    done()          // terminates this coordinator cleanly
+}
+```
+
+Internally, `done()` throws `KlavaNodeDoneException` (an unchecked exception).
+`KlavaNodeCoordinator.execute()` catches it as normal completion — no stack trace,
+no error log. Any code after `done()` is unreachable.
+
+Without a `done()` call the coordinator keeps running indefinitely (waiting for
+further tuple-space operations), which means `waitForCompletion()` in `main()` would
+block forever.
+
+---
+
+## `waitForCompletion()`
+
+`KlavaNode.waitForCompletion()` performs the following steps in order:
+
+1. **Join the main coordinator thread** (if one was registered via `setMainCoordinator()`).
+   This blocks until `executeProcess()` returns — either normally, via `done()`, or via
+   an interrupt.
+2. **Recursively call `waitForCompletion()`** on every node registered via
+   `addManagedNode()`. For a net, these are the nodes created by `addNodes()`.
+3. **Call `close()`** to release all sockets, sessions, and resources.
+
+A timeout variant is also available:
+
+```java
+node.waitForCompletion(5000); // wait at most 5 000 ms
+```
+
+---
+
+## `addMainProcess()` and `setMainCoordinator()`
+
+`addMainProcess()` always does two distinct things:
+
+```java
+KlavaNodeCoordinator _coordinator = new <NodeName>Process();
+setMainCoordinator(_coordinator);   // tells waitForCompletion() which thread to join
+addNodeCoordinator(_coordinator);   // registers and starts the coordinator thread
+```
+
+Both calls are necessary:
+
+- `addNodeCoordinator` starts the process running in a thread managed by the node.
+- `setMainCoordinator` stores a reference to that thread so `waitForCompletion()` can
+  join it later.
+
+---
+
+## Lifecycle diagrams
+
+### Standalone node
+
+```
+new NodeClass()
+└─ setupEnvironment()          (if [env] block present)
+└─ addMainProcess()
+   ├─ setMainCoordinator(…)
+   └─ addNodeCoordinator(…)    ← coordinator thread starts here
+      └─ executeProcess()      ← runs DSL body; calls done() when finished
+└─ waitForCompletion()
+   ├─ join(mainCoordinator)    ← blocks until executeProcess() returns
+   └─ close()
+```
+
+### Net with nodes
+
+```
+new NetClass()
+└─ addNodes()
+   ├─ new NodeA(); addManagedNode(nodeA); nodeA.addMainProcess()
+   └─ new NodeB(); addManagedNode(nodeB); nodeB.addMainProcess()
+└─ waitForCompletion()
+   ├─ nodeA.waitForCompletion()
+   │  ├─ join(nodeA.mainCoordinator)
+   │  └─ nodeA.close()
+   ├─ nodeB.waitForCompletion()
+   │  ├─ join(nodeB.mainCoordinator)
+   │  └─ nodeB.close()
+   └─ net.close()
+```
+
+---
+
+## Shutdown noise suppression
+
+When a node closes, the transport layer produces `EOFException` (remote side closed)
+and `SocketException("Socket closed")` (local side closed). These are normal and are
+silently downgraded to DEBUG level by `ProtocolExceptionUtils.isCausedByConnectionClose()`,
+which traverses both the standard Java cause chain (`Throwable.getCause()`) and
+`ProtocolException`'s custom `actual` field (via `ProtocolException.represents()`).

--- a/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/protocols/ProtocolExceptionUtils.java
+++ b/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/protocols/ProtocolExceptionUtils.java
@@ -1,0 +1,46 @@
+package org.mikado.imc.protocols;
+
+import java.io.EOFException;
+import java.net.SocketException;
+
+/**
+ * Utility methods for inspecting exception cause chains that include
+ * {@link ProtocolException}, which stores its wrapped exception in a custom
+ * field rather than the standard Java cause chain.
+ * 
+ * @author Lorenzo Bettini
+ */
+public class ProtocolExceptionUtils {
+
+    private ProtocolExceptionUtils() {
+        // utility class
+    }
+
+    /**
+     * Returns {@code true} if the given throwable, or any exception in its
+     * cause chain (including inside {@link ProtocolException} wrappers), is an
+     * {@link EOFException} or a {@link SocketException} with the message
+     * {@code "Socket closed"}.
+     *
+     * <p>Both of these indicate a normal connection close rather than an
+     * unexpected protocol error.
+     *
+     * @param t the throwable to inspect
+     * @return {@code true} if caused by a connection-close condition
+     */
+    public static boolean isCausedByConnectionClose(Throwable t) {
+        while (t != null) {
+            if (t instanceof EOFException)
+                return true;
+            if (t instanceof SocketException
+                    && "Socket closed".equals(t.getMessage()))
+                return true;
+            if (t instanceof ProtocolException) {
+                t = ((ProtocolException) t).represents();
+            } else {
+                t = t.getCause();
+            }
+        }
+        return false;
+    }
+}

--- a/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/protocols/ProtocolStack.java
+++ b/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/protocols/ProtocolStack.java
@@ -145,7 +145,7 @@ public class ProtocolStack {
             throws ProtocolException {
         UnMarshaler unMarshaler = null;
 
-        if (layers.size() == 0) {
+        if (layers.isEmpty()) {
             throw new ProtocolException("protocol stack is empty");
         }
 

--- a/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
+++ b/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
@@ -46,7 +46,11 @@ public class ProtocolThread extends NodeProcess {
         try {
             protocol.start();
         } catch (ProtocolException e) {
-            LOGGER.error("protocol error in thread {}", getName(), e);
+            if (isCausedByConnectionClose(e)) {
+                LOGGER.debug("connection closed in thread {}", getName());
+            } else {
+                LOGGER.error("protocol error in thread {}", getName(), e);
+            }
         } finally {
             try {
                 protocol.close();
@@ -61,6 +65,18 @@ public class ProtocolThread extends NodeProcess {
      * 
      * @see org.mikado.imc.topology.NodeProcess#close()
      */
+    private static boolean isCausedByConnectionClose(Throwable t) {
+        while (t != null) {
+            if (t instanceof java.io.EOFException)
+                return true;
+            if (t instanceof java.net.SocketException
+                    && "Socket closed".equals(t.getMessage()))
+                return true;
+            t = t.getCause();
+        }
+        return false;
+    }
+
     @Override
     public void close() throws IMCException {
         super.close();

--- a/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
+++ b/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
@@ -7,6 +7,7 @@ package org.mikado.imc.topology;
 import org.mikado.imc.common.IMCException;
 import org.mikado.imc.protocols.Protocol;
 import org.mikado.imc.protocols.ProtocolException;
+import org.mikado.imc.protocols.ProtocolExceptionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -46,7 +47,7 @@ public class ProtocolThread extends NodeProcess {
         try {
             protocol.start();
         } catch (ProtocolException e) {
-            if (isCausedByConnectionClose(e)) {
+            if (ProtocolExceptionUtils.isCausedByConnectionClose(e)) {
                 LOGGER.debug("connection closed in thread {}", getName());
             } else {
                 LOGGER.error("protocol error in thread {}", getName(), e);
@@ -65,22 +66,6 @@ public class ProtocolThread extends NodeProcess {
      * 
      * @see org.mikado.imc.topology.NodeProcess#close()
      */
-    private static boolean isCausedByConnectionClose(Throwable t) {
-        while (t != null) {
-            if (t instanceof java.io.EOFException)
-                return true;
-            if (t instanceof java.net.SocketException
-                    && "Socket closed".equals(t.getMessage()))
-                return true;
-            if (t instanceof ProtocolException) {
-                t = ((ProtocolException) t).represents();
-            } else {
-                t = t.getCause();
-            }
-        }
-        return false;
-    }
-
     @Override
     public void close() throws IMCException {
         super.close();

--- a/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
+++ b/xklaim/xklaim.parent/imc/src/main/java/org/mikado/imc/topology/ProtocolThread.java
@@ -72,7 +72,11 @@ public class ProtocolThread extends NodeProcess {
             if (t instanceof java.net.SocketException
                     && "Socket closed".equals(t.getMessage()))
                 return true;
-            t = t.getCause();
+            if (t instanceof ProtocolException) {
+                t = ((ProtocolException) t).represents();
+            } else {
+                t = t.getCause();
+            }
         }
         return false;
     }

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
@@ -3,6 +3,7 @@
  */
 package klava.proto;
 
+import java.io.EOFException;
 import java.io.IOException;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -225,11 +226,24 @@ public class MessageState extends ProtocolStateSimple {
                     .enter(null, new TransmissionChannel(unMarshaler));
         } catch (ProtocolException e) {
             printSession();
-            LOGGER.error("protocol error in message state", e);
+            if (isCausedByEOF(e)) {
+                LOGGER.debug("connection closed in message state");
+            } else {
+                LOGGER.error("protocol error in message state", e);
+            }
             lostConnection();
             throw e;
             /* something bad happened so we end the protocol */
         }
+    }
+
+    private static boolean isCausedByEOF(Throwable t) {
+        while (t != null) {
+            if (t instanceof EOFException)
+                return true;
+            t = t.getCause();
+        }
+        return false;
     }
 
     /**

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
@@ -244,7 +244,11 @@ public class MessageState extends ProtocolStateSimple {
             if (t instanceof java.net.SocketException
                     && "Socket closed".equals(t.getMessage()))
                 return true;
-            t = t.getCause();
+            if (t instanceof ProtocolException) {
+                t = ((ProtocolException) t).represents();
+            } else {
+                t = t.getCause();
+            }
         }
         return false;
     }

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
@@ -3,7 +3,6 @@
  */
 package klava.proto;
 
-import java.io.EOFException;
 import java.io.IOException;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -19,6 +18,7 @@ import org.mikado.imc.common.IMCException;
 import org.mikado.imc.mobility.MigratingCodeFactory;
 import org.mikado.imc.protocols.Marshaler;
 import org.mikado.imc.protocols.ProtocolException;
+import org.mikado.imc.protocols.ProtocolExceptionUtils;
 import org.mikado.imc.protocols.ProtocolStack;
 import org.mikado.imc.protocols.ProtocolStateSimple;
 import org.mikado.imc.protocols.ProtocolSwitchState;
@@ -226,7 +226,7 @@ public class MessageState extends ProtocolStateSimple {
                     .enter(null, new TransmissionChannel(unMarshaler));
         } catch (ProtocolException e) {
             printSession();
-            if (isCausedByConnectionClose(e)) {
+            if (ProtocolExceptionUtils.isCausedByConnectionClose(e)) {
                 LOGGER.debug("connection closed in message state");
             } else {
                 LOGGER.error("protocol error in message state", e);
@@ -235,22 +235,6 @@ public class MessageState extends ProtocolStateSimple {
             throw e;
             /* something bad happened so we end the protocol */
         }
-    }
-
-    private static boolean isCausedByConnectionClose(Throwable t) {
-        while (t != null) {
-            if (t instanceof EOFException)
-                return true;
-            if (t instanceof java.net.SocketException
-                    && "Socket closed".equals(t.getMessage()))
-                return true;
-            if (t instanceof ProtocolException) {
-                t = ((ProtocolException) t).represents();
-            } else {
-                t = t.getCause();
-            }
-        }
-        return false;
     }
 
     /**

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/proto/MessageState.java
@@ -226,7 +226,7 @@ public class MessageState extends ProtocolStateSimple {
                     .enter(null, new TransmissionChannel(unMarshaler));
         } catch (ProtocolException e) {
             printSession();
-            if (isCausedByEOF(e)) {
+            if (isCausedByConnectionClose(e)) {
                 LOGGER.debug("connection closed in message state");
             } else {
                 LOGGER.error("protocol error in message state", e);
@@ -237,9 +237,12 @@ public class MessageState extends ProtocolStateSimple {
         }
     }
 
-    private static boolean isCausedByEOF(Throwable t) {
+    private static boolean isCausedByConnectionClose(Throwable t) {
         while (t != null) {
             if (t instanceof EOFException)
+                return true;
+            if (t instanceof java.net.SocketException
+                    && "Socket closed".equals(t.getMessage()))
                 return true;
             t = t.getCause();
         }

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
@@ -3,7 +3,6 @@
  */
 package klava.topology;
 
-import java.net.SocketException;
 import java.util.List;
 
 import klava.KlavaException;
@@ -13,7 +12,7 @@ import klava.PhysicalLocality;
 import klava.Tuple;
 
 import org.mikado.imc.common.IMCException;
-import org.mikado.imc.protocols.ProtocolException;
+import org.mikado.imc.protocols.ProtocolExceptionUtils;
 import org.mikado.imc.protocols.ProtocolStack;
 import org.mikado.imc.topology.NodeCoordinator;
 import org.mikado.imc.topology.NodeCoordinatorProxy;
@@ -91,7 +90,7 @@ public abstract class KlavaNodeCoordinator extends NodeCoordinator {
                 Thread.currentThread().interrupt();
                 return;
             }
-            if (isCausedBySocketClosed(e)) {
+            if (ProtocolExceptionUtils.isCausedByConnectionClose(e)) {
                 LOGGER.debug("connection closed in coordinator {}", getName());
                 return;
             }
@@ -99,20 +98,6 @@ public abstract class KlavaNodeCoordinator extends NodeCoordinator {
             throw new IMCException("Uncaught exception", e);
         }
 
-    }
-
-    private static boolean isCausedBySocketClosed(Throwable t) {
-        while (t != null) {
-            if (t instanceof SocketException
-                    && "Socket closed".equals(t.getMessage()))
-                return true;
-            if (t instanceof ProtocolException) {
-                t = ((ProtocolException) t).represents();
-            } else {
-                t = t.getCause();
-            }
-        }
-        return false;
     }
 
     /**

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
@@ -3,6 +3,7 @@
  */
 package klava.topology;
 
+import java.net.SocketException;
 import java.util.List;
 
 import klava.KlavaException;
@@ -89,11 +90,24 @@ public abstract class KlavaNodeCoordinator extends NodeCoordinator {
                 Thread.currentThread().interrupt();
                 return;
             }
-
+            if (isCausedBySocketClosed(e)) {
+                LOGGER.debug("connection closed in coordinator {}", getName());
+                return;
+            }
             LOGGER.error("uncaught exception in coordinator {}", getName(), e);
             throw new IMCException("Uncaught exception", e);
         }
 
+    }
+
+    private static boolean isCausedBySocketClosed(Throwable t) {
+        while (t != null) {
+            if (t instanceof SocketException
+                    && "Socket closed".equals(t.getMessage()))
+                return true;
+            t = t.getCause();
+        }
+        return false;
     }
 
     /**

--- a/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
+++ b/xklaim/xklaim.parent/klava/src/main/java/klava/topology/KlavaNodeCoordinator.java
@@ -13,6 +13,7 @@ import klava.PhysicalLocality;
 import klava.Tuple;
 
 import org.mikado.imc.common.IMCException;
+import org.mikado.imc.protocols.ProtocolException;
 import org.mikado.imc.protocols.ProtocolStack;
 import org.mikado.imc.topology.NodeCoordinator;
 import org.mikado.imc.topology.NodeCoordinatorProxy;
@@ -105,7 +106,11 @@ public abstract class KlavaNodeCoordinator extends NodeCoordinator {
             if (t instanceof SocketException
                     && "Socket closed".equals(t.getMessage()))
                 return true;
-            t = t.getCause();
+            if (t instanceof ProtocolException) {
+                t = ((ProtocolException) t).represents();
+            } else {
+                t = t.getCause();
+            }
         }
         return false;
     }

--- a/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelection/LeaderElection.java
+++ b/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelection/LeaderElection.java
@@ -5,5 +5,6 @@ public class LeaderElection {
   public static void main(final String[] args) throws Exception {
     LeaderElectionNet leaderElectionNet = new LeaderElectionNet();
     leaderElectionNet.addNodes();
+    leaderElectionNet.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelection/LeaderElectionNet.java
+++ b/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelection/LeaderElectionNet.java
@@ -43,7 +43,9 @@ public class LeaderElectionNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionNet.L1.L1Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionNet.L1.L1Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -67,7 +69,9 @@ public class LeaderElectionNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionNet.L2.L2Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionNet.L2.L2Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -91,7 +95,9 @@ public class LeaderElectionNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionNet.L3.L3Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionNet.L3.L3Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -110,7 +116,9 @@ public class LeaderElectionNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionNet.RG.RGProcess());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionNet.RG.RGProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -126,6 +134,10 @@ public class LeaderElectionNet extends LogicalNet {
     l1.setupEnvironment();
     l2.setupEnvironment();
     l3.setupEnvironment();
+    addManagedNode(l1);
+    addManagedNode(l2);
+    addManagedNode(l3);
+    addManagedNode(rG);
     l1.addMainProcess();
     l2.addMainProcess();
     l3.addMainProcess();

--- a/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelectionrec/LeaderElectionRecursive.java
+++ b/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelectionrec/LeaderElectionRecursive.java
@@ -5,5 +5,6 @@ public class LeaderElectionRecursive {
   public static void main(final String[] args) throws Exception {
     LeaderElectionRecursiveNet leaderElectionRecursiveNet = new LeaderElectionRecursiveNet();
     leaderElectionRecursiveNet.addNodes();
+    leaderElectionRecursiveNet.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelectionrec/LeaderElectionRecursiveNet.java
+++ b/xklaim/xklaim.parent/xklaim.example.leaderelection/src-gen/xklaim/example/leaderelectionrec/LeaderElectionRecursiveNet.java
@@ -46,7 +46,9 @@ public class LeaderElectionRecursiveNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionRecursiveNet.L1.L1Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionRecursiveNet.L1.L1Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -70,7 +72,9 @@ public class LeaderElectionRecursiveNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionRecursiveNet.L2.L2Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionRecursiveNet.L2.L2Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -94,7 +98,9 @@ public class LeaderElectionRecursiveNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionRecursiveNet.L3.L3Process());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionRecursiveNet.L3.L3Process();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -113,7 +119,9 @@ public class LeaderElectionRecursiveNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new LeaderElectionRecursiveNet.RG.RGProcess());
+      KlavaNodeCoordinator _coordinator = new LeaderElectionRecursiveNet.RG.RGProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -129,6 +137,10 @@ public class LeaderElectionRecursiveNet extends LogicalNet {
     l1.setupEnvironment();
     l2.setupEnvironment();
     l3.setupEnvironment();
+    addManagedNode(l1);
+    addManagedNode(l2);
+    addManagedNode(l3);
+    addManagedNode(rG);
     l1.addMainProcess();
     l2.addMainProcess();
     l3.addMainProcess();

--- a/xklaim/xklaim.parent/xklaim.example.mobility.receiver/src-gen/xklaim/example/mobility/receiver/CodeMobilityReceiver.java
+++ b/xklaim/xklaim.parent/xklaim.example.mobility.receiver/src-gen/xklaim/example/mobility/receiver/CodeMobilityReceiver.java
@@ -5,5 +5,6 @@ public class CodeMobilityReceiver {
   public static void main(final String[] args) throws Exception {
     Receiver receiver = new Receiver();
     receiver.addMainProcess();
+    receiver.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.example.mobility.receiver/src-gen/xklaim/example/mobility/receiver/Receiver.java
+++ b/xklaim/xklaim.parent/xklaim.example.mobility.receiver/src-gen/xklaim/example/mobility/receiver/Receiver.java
@@ -32,6 +32,8 @@ public class Receiver extends KlavaNode {
   }
 
   public void addMainProcess() throws IMCException {
-    addNodeCoordinator(new Receiver.ReceiverProcess());
+    KlavaNodeCoordinator _coordinator = new Receiver.ReceiverProcess();
+    setMainCoordinator(_coordinator);
+    addNodeCoordinator(_coordinator);
   }
 }

--- a/xklaim/xklaim.parent/xklaim.example.mobility.sender/src-gen/xklaim/example/mobility/sender/CodeMobilitySender.java
+++ b/xklaim/xklaim.parent/xklaim.example.mobility.sender/src-gen/xklaim/example/mobility/sender/CodeMobilitySender.java
@@ -6,5 +6,6 @@ public class CodeMobilitySender {
     Sender sender = new Sender();
     sender.setupEnvironment();
     sender.addMainProcess();
+    sender.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.example.mobility.sender/src-gen/xklaim/example/mobility/sender/Sender.java
+++ b/xklaim/xklaim.parent/xklaim.example.mobility.sender/src-gen/xklaim/example/mobility/sender/Sender.java
@@ -54,6 +54,8 @@ public class Sender extends KlavaNode {
   }
 
   public void addMainProcess() throws IMCException {
-    addNodeCoordinator(new Sender.SenderProcess());
+    KlavaNodeCoordinator _coordinator = new Sender.SenderProcess();
+    setMainCoordinator(_coordinator);
+    addNodeCoordinator(_coordinator);
   }
 }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/hello/Hello.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/hello/Hello.java
@@ -5,5 +5,6 @@ public class Hello {
   public static void main(final String[] args) throws Exception {
     HelloNet helloNet = new HelloNet();
     helloNet.addNodes();
+    helloNet.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/hello/HelloNet.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/hello/HelloNet.java
@@ -48,7 +48,9 @@ public class HelloNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloNet.Reader.ReaderProcess());
+      KlavaNodeCoordinator _coordinator = new HelloNet.Reader.ReaderProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -65,7 +67,9 @@ public class HelloNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloNet.Writer.WriterProcess());
+      KlavaNodeCoordinator _coordinator = new HelloNet.Writer.WriterProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -77,6 +81,8 @@ public class HelloNet extends LogicalNet {
     HelloNet.Reader reader = new HelloNet.Reader();
     HelloNet.Writer writer = new HelloNet.Writer();
     reader.setupEnvironment();
+    addManagedNode(reader);
+    addManagedNode(writer);
     reader.addMainProcess();
     writer.addMainProcess();
   }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremote/HelloFromReceivedProc.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremote/HelloFromReceivedProc.java
@@ -5,5 +5,6 @@ public class HelloFromReceivedProc {
   public static void main(final String[] args) throws Exception {
     HelloFromReceivedProcNet helloFromReceivedProcNet = new HelloFromReceivedProcNet();
     helloFromReceivedProcNet.addNodes();
+    helloFromReceivedProcNet.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremote/HelloFromReceivedProcNet.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremote/HelloFromReceivedProcNet.java
@@ -54,7 +54,9 @@ public class HelloFromReceivedProcNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloFromReceivedProcNet.Reader.ReaderProcess());
+      KlavaNodeCoordinator _coordinator = new HelloFromReceivedProcNet.Reader.ReaderProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -80,7 +82,9 @@ public class HelloFromReceivedProcNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloFromReceivedProcNet.Writer.WriterProcess());
+      KlavaNodeCoordinator _coordinator = new HelloFromReceivedProcNet.Writer.WriterProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -91,6 +95,8 @@ public class HelloFromReceivedProcNet extends LogicalNet {
   public void addNodes() throws IMCException {
     HelloFromReceivedProcNet.Reader reader = new HelloFromReceivedProcNet.Reader();
     HelloFromReceivedProcNet.Writer writer = new HelloFromReceivedProcNet.Writer();
+    addManagedNode(reader);
+    addManagedNode(writer);
     reader.addMainProcess();
     writer.addMainProcess();
   }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremoteeval/HelloRemoveEvalProc.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremoteeval/HelloRemoveEvalProc.java
@@ -5,5 +5,6 @@ public class HelloRemoveEvalProc {
   public static void main(final String[] args) throws Exception {
     HelloRemoveEvalProcNet helloRemoveEvalProcNet = new HelloRemoveEvalProcNet();
     helloRemoveEvalProcNet.addNodes();
+    helloRemoveEvalProcNet.waitForCompletion();
   }
 }

--- a/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremoteeval/HelloRemoveEvalProcNet.java
+++ b/xklaim/xklaim.parent/xklaim.examples/src-gen/xklaim/examples/helloremoteeval/HelloRemoveEvalProcNet.java
@@ -56,7 +56,9 @@ public class HelloRemoveEvalProcNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloRemoveEvalProcNet.Reader.ReaderProcess());
+      KlavaNodeCoordinator _coordinator = new HelloRemoveEvalProcNet.Reader.ReaderProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -73,7 +75,9 @@ public class HelloRemoveEvalProcNet extends LogicalNet {
     }
 
     public void addMainProcess() throws IMCException {
-      addNodeCoordinator(new HelloRemoveEvalProcNet.Writer.WriterProcess());
+      KlavaNodeCoordinator _coordinator = new HelloRemoveEvalProcNet.Writer.WriterProcess();
+      setMainCoordinator(_coordinator);
+      addNodeCoordinator(_coordinator);
     }
   }
 
@@ -84,6 +88,8 @@ public class HelloRemoveEvalProcNet extends LogicalNet {
   public void addNodes() throws IMCException {
     HelloRemoveEvalProcNet.Reader reader = new HelloRemoveEvalProcNet.Reader();
     HelloRemoveEvalProcNet.Writer writer = new HelloRemoveEvalProcNet.Writer();
+    addManagedNode(reader);
+    addManagedNode(writer);
     reader.addMainProcess();
     writer.addMainProcess();
   }

